File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/count-of-substrings-containing-every-vowel-and-k-consonants-ii.py
class Solution:
    def countOfSubstrings(self, word: str, k: int) -> int:
        n = len(word)
        vowels = 'aeiou'
        vowset = set(vowels)
        vowCnt = defaultdict(int)

        cons = 0
        left = count = substr = 0

        def minus(ch):
            if ch in vowset:
                vowCnt[ch] -= 1
                if vowCnt[ch] == 0:
                    del vowCnt[ch]
            else:
                nonlocal cons
                cons -= 1
        
        for ch in word:
            if ch in vowset:
                vowCnt[ch] += 1
            else:
                cons += 1
                count = 0
            
            while cons > k:
                minus(word[left])
                left += 1
            while cons == k and len(vowCnt) == 5:
                count += 1
                minus(word[left])
                left += 1
            substr += count
        
        return substr

File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/count-of-substrings-containing-every-vowel-and-k-consonants-ii.java
class Solution {
    public long countOfSubstrings(String word, int k) {
        int n = word.length();
        String vowels = "aeiou";
        Set<Character> vs = new HashSet<>();

        for (char ch : vowels.toCharArray()) {
            vs.add(ch);
        }

        // map to count occurences of vowels in curr window
        Map<Character, Integer> vcount = new HashMap<>();
        long count = 0;
        int left = 0;
        long substr = 0;
        long cons = 0;

        for(int i = 0; i < n; i++) {
            char ch = word.charAt(i);
            if (vs.contains(ch)) {
                vcount.put(ch, vcount.getOrDefault(ch, 0) + 1);
            } else {
                cons++;
                count = 0; // we have to reset count when cons is added
            }

            while(cons > k) {
                char leftch = word.charAt(left);
                if(vs.contains(leftch)) {
                    int curr = vcount.get(leftch);
                    if(curr == 1) {
                        vcount.remove(leftch);
                    } else {
                        vcount.put(leftch, curr - 1);
                    }
                } else {
                    cons--;
                }
                left++;
            }

            while (cons == k && vcount.size() == 5) {
                count++;
                char leftch = word.charAt(left);
                if (vs.contains(leftch)) {
                    int curr = vcount.get(leftch);
                    if (curr == 1) {
                        vcount.remove(leftch);
                    } else {
                        vcount.put(leftch, curr - 1);
                    }
                } else {
                    cons--;
                }
                left++;
            }
            
            // Add the count for this window to the overall substring count.
            substr += count;
        }
        return substr;
    }
}